src/signature.rs

```diff
@@ -0,0 +1,71 @@
+//! Device electronic signature
+//!
+//! (stored in flash memory)
+
+use core::str::from_utf8_unchecked;
+
+/// Uniqure Device ID register
+#[derive(Hash, Debug)]
+#[repr(C)]
+pub struct Uid {
+    x: u16,
+    y: u16,
+    waf_lot: [u8; 8],
+}
+
+impl Uid {
+    fn ptr() -> *const Self {
+        0x1FFF_7A10 as *const _
+    }
+
+    /// Returns a wrapped reference to the value in flash memory
+    pub fn get() -> &'static Self {
+        unsafe { &*Self::ptr() }
+    }
+
+    /// X coordinate on wafer
+    pub fn x(&self) -> u16 {
+        self.x
+    }
+
+    /// Y coordinate on wafer
+    pub fn y(&self) -> u16 {
+        self.y
+    }
+
+    /// Wafer number
+    pub fn waf_num(&self) -> u8 {
+        self.waf_lot[0].into()
+    }
+
+    /// Lot number
+    pub unsafe fn lot_num(&self) -> &str {
+        from_utf8_unchecked(&self.waf_lot[1..])
+    }
```

Comment:
Any issue with putting the from_utf8_unchecked() inside an unsafe block rather than making the function unsafe ?

Reply:
Is ASCII from flash potentially unsafe (corrupted) in the real world?

Reply:
I have no idea. Ideally we would check with as much different STM32 as possible, but I guess for now we should just trust the ST Reference Manual which specify it is ASCII. If we encounter some example with non-ASCII in the future, we will think about a solution then.